# NOPT042 Constraint programming: Tutorial 4 – Search strategies

### From last week:

* Solution to the Coin grid problem.
* Best model and solver for the problem? MIP, naturally expressed as an integer program
* Unsatisfiable instances - LP works well.
* For sparse solution sets heuristic approaches may be slow.

## Example: N-queens

Place $n$ queens on an $n\times n$ board so that none attack another. How to choose the decision variables?

* How large is the search space?
* Can we use symmetry breaking?
* Consider the _dual_ model.

In [1]:
!time picat queens/queens-primal.pi 8

.....Q..
...Q....
.Q......
.......Q
....Q...
......Q.
Q.......
..Q.....

real	0m0.034s
user	0m0.010s
sys	0m0.017s


In [2]:
!cat queens/queens-primal.pi

% n-queens, primal model
import sat.

main([N]) =>
    N := to_int(N),
    queens(N, Q),
    solve(Q),
    if N <= 32 then
        output(Q)
    end.


queens(N, Q) =>
    Q = new_array(N),
    Q :: 1..N,
    all_different(Q),
    all_different([$Q[I] - I : I in 1..N]),
    all_different([$Q[I] + I : I in 1..N]).


output(Q) =>
    N = Q.length,
    foreach(I in 1..N)
        foreach (J in 1..N)
            if Q[I] = J then
                print("Q")
            else
                print(".")
            end
        end,
        print("\n")
    end.


In [3]:
!time picat queens/queens-dual.pi 64

Welcome to the CBC MILP Solver 
Version: 2.10.7 
Build Date: Feb 14 2022 

command line - cbc __tmp.lp solve solu __tmp.sol (default strategy 1)
Continuous objective value is 0 - 0.13 seconds
Cgl0003I 0 fixed, 0 tightened bounds, 2 strengthened rows, 0 substitutions
Cgl0004I processed model has 379 rows, 4096 columns (4096 integer (4096 of which binary)) and 20478 elements
Cbc0045I No integer variables out of 4096 objects (4096 integer) have costs
Cbc0045I branch on satisfied N create fake objective Y random cost Y
Cbc0038I Initial state - 125 integers unsatisfied sum - 26.3403
Cbc0038I Pass   1: suminf.   10.43351 (69) obj. 0 iterations 767
Cbc0038I Pass   2: suminf.    6.76111 (160) obj. 0 iterations 471
Cbc0038I Pass   3: suminf.    5.28571 (23) obj. 0 iterations 743
Cbc0038I Pass   4: suminf.    5.24859 (154) obj. 0 iterations 160
Cbc0038I Pass   5: suminf.    2.00000 (4) obj. 0 iterations 379
Cbc0038I Pass   6: suminf.    2.00000 (4) obj. 0 iterations 139
Cbc0038I Pass   7: suminf

In [4]:
!cat queens/queens-dual.pi

% n-queens, dual model
import mip.

main([N]) =>
    N := to_int(N),
    queens(N, Board),
    solve(Board),
    if N <= 32 then
        output(Q)
    end.


queens(N, Board) =>
    Board = new_array(N, N),
    Board :: 0..1,
    
    sum([Board[I, J] : I in 1..N, J in 1..N]) #= N,
    
    % rows
    foreach(I in 1..N)
        sum([Board[I, J] : J in 1..N]) #<= 1
    end,
    
    % cols
    foreach(J in 1..N)
        sum([Board[I, J] : I in 1..N]) #<= 1
    end,

    % diags
    foreach(K in 1-N..N-1)
        sum([Board[I,J] : I in 1..N, J in 1..N, I-J = K ]) #<= 1
    end,
    foreach(K in 2..2*N)
        sum([Board[I,J] : I in 1..N, J in 1..N, I+J = K ]) #<= 1
    end.
    

output(Board) =>
    N = Board.length,
    foreach(I in 1..N)
        foreach (J in 1..N)
            if Board[I, J] = 1 then
                print("Q")
            else
                print(".")
            end
        end,
        print("\n")
    end.


Sometimes it is best to model the problem in both ways and add _channelling constraints_. (Here it does not help.)

In [5]:
!time picat queens/queens-channeling.pi 128


real	0m3.502s
user	0m3.353s
sys	0m0.141s


In [6]:
!cat queens/queens-channeling.pi

% n-queens, primal model
import sat.

main([N]) =>
    N := to_int(N),
    queens(N, Q, Board),
    solve(Q ++ Board),
    if N <= 32 then   
        output(Q)
    end.


queens(N, Q, Board) =>    
    % primal
    Q = new_array(N),
    Q :: 1..N,
    all_different(Q),
    all_different([$Q[I] - I : I in 1..N]),
    all_different([$Q[I] + I : I in 1..N]),

    % dual
    Board = new_array(N, N),
    Board :: 0..1,    
    sum([Board[I, J] : I in 1..N, J in 1..N]) #= N,
    foreach(I in 1..N)
        sum([Board[I, J] : J in 1..N]) #<= 1
    end,
    foreach(J in 1..N)
        sum([Board[I, J] : I in 1..N]) #<= 1
    end,
    foreach(K in 1-N..N-1)
        sum([Board[I, J] : I in 1..N, J in 1..N, I - J = K ]) #<= 1
    end,
    foreach(K in 2..2*N)
        sum([Board[I, J] : I in 1..N, J in 1..N, I + J = K ]) #<= 1
    end,
    
    % channeling
    foreach(I in 1..N, J in 1..N)
        (Board[I,J] #= 1) #<=> (Q[I] #= J)
    end.


output(Q) =>
    N = Q.length,
    foreach(I in 1..N)
  

Can the models be improved using symmetry breaking?

## Search strategies
And other solver options: see [Picat guide](http://picat-lang.org/download/picat_guide.pdf) (Section 12.6) and the [book](http://picat-lang.org/picatbook2015/constraint_solving_and_planning_with_picat.pdf) (Section 3.5)

In [7]:
%load_ext ipicat

<IPython.core.display.Javascript object>

Picat version 3.5#5


In [8]:
%%picat -n queens
import cp. %try sat, try also mip with the other model

queens(N, Q) =>
    Q = new_array(N),
    Q :: 1..N,
    all_different(Q),
    all_different([$Q[I] - I : I in 1..N]),
    all_different([$Q[I] + I : I in 1..N]).

The predicate `time2` also outputs the number of backtracks during the search - a good measure of complexity.

In [9]:
%%picat
main =>
    N = 32,
    queens(N, Q),
    time2(solve(Q)).


CPU time 28.687 seconds. Backtracks: 11461548




Which search strategy could work well for our model?

Here's how we can test multiple search strategies (code adapted from [the book](https://picat-lang.org/picatbook2015/constraint_solving_and_planning_with_picat.pdf)):

In [10]:
%%picat

% Variable selection
selection(VarSels) =>
    VarSels = [backward,constr,degree,ff,ffc,ffd,forward,inout,leftmost,max,min,up].
    
% Value selection
choice(ValSels) =>
    ValSels = [down,reverse_split,split,up,updown].

main =>
    selection(VarSels),
    choice(ValSels),
    Timeout = 1000, % Timeout in milliseconds
    %Timeout = 10000, % Timeout in milliseconds
    Ns = [64, 128, 256],
    
    foreach (N in Ns, VarSel in VarSels, ValSel in ValSels)
        queens(N,Q),
        time2(time_out(solve([VarSel,ValSel], Q),Timeout,Status)),
        println([N,VarSel,ValSel,Status])
    end.


CPU time 0.988 seconds. Backtracks: 345745

[64,backward,down,time_out]

CPU time 0.973 seconds. Backtracks: 0

[64,backward,reverse_split,time_out]

CPU time 0.992 seconds. Backtracks: 0

[64,backward,split,time_out]

CPU time 0.997 seconds. Backtracks: 224513

[64,backward,up,time_out]

CPU time 0.99 seconds. Backtracks: 498793

[64,backward,updown,time_out]

CPU time 0.998 seconds. Backtracks: 363610

[64,constr,down,time_out]

CPU time 0.998 seconds. Backtracks: 0

[64,constr,reverse_split,time_out]

CPU time 0.995 seconds. Backtracks: 0

[64,constr,split,time_out]

CPU time 0.992 seconds. Backtracks: 212687

[64,constr,up,time_out]

CPU time 0.996 seconds. Backtracks: 508551

[64,constr,updown,time_out]

CPU time 0.995 seconds. Backtracks: 355107

[64,degree,down,time_out]

CPU time 0.994 seconds. Backtracks: 0

[64,degree,reverse_split,time_out]

CPU time 0.997 seconds. Backtracks: 0

[64,degree,split,time_out]

CPU time 0.988 seconds. Backtracks: 198952

[64,degree,up,time_out]

## Exercises

### Exercise: Magic square
Arrange numbers $1,2,\dots,n^2$ in a square such that every row, every column, and the two main diagonals all sum to the same quantity. 

* Try to find the best model, solver and search strategy.
* How many magic squares are there for a given $n$? 
* Allow also for a partially filled instance. 

### Exercise: Minesweeper
Identify the positions of all mines in a given board. Try the following instance (from [the book](http://picat-lang.org/picatbook2015/constraint_solving_and_planning_with_picat.pdf)):
```
Instance = {
    {_,_,2,_,3,_},
    {2,_,_,_,_,_},
    {_,_,2,4,_,3},
    {1,_,3,4,_,_},
    {_,_,_,_,_,3},
    {_,3,_,3,_,_}
}.
```

### Exercise: Graph-coloring
1. Write a program that solves the (directed) graph 3-coloring problem with a given number of colors and a given graph. The graph is given by a list of edges, each edge is a 2-element list. We assume that vertices of the graph are $1,\dots,n$ where $n$ is the maximum number appearing in the list. 
2. Generalize your program to graph $k$-coloring where $k$ is a positive integer given on the input.
3. Modify your program to accept the incidence matrix (a 2D array) instead of the list of edges.
4. Add the flag `-n` to output the minimum number of colors (the chromatic number) of a given graph.
For example:
```
picat graph-coloring.pi [[1,2],[2,3],[3,4],[4,1]]
picat graph-coloring.pi [[1,2],[2,3],[3,1]] 4
picat graph-coloring.pi "{{0,1,1},{1,0,1},{1,1,0}}" 4
picat graph-coloring.pi -n [[1,2],[2,3],[3,4],[4,1]]
```

# Knapsack

There are two common versions of the problem: the general __knapsack__ problem:

> Given a set of items, each with a weight and a value, determine __how many of each item__ to include in a collection so that the total weight is less than or equal to a given limit and the total value is as large as possible.

And the __0-1 knapsack__ problem:


> Given a set of items, each with a weight and a value, determine __which items__ to include in a collection so that the total weight is less than or equal to a given limit and the total value is as large as possible.

(In a general knapsack problem, we can take any number of each item, in the 0-1 version we can take at most one of each.)

#### Example of an instance:

A thief breaks into a department store (general knapsack) or into a home (0-1 knapsack). They can carry 23kg. Which items (and how many of each, in the general version) should they take to maximize profit? There are the following items:

* a TV (weighs 15kg, costs \$500),
* a desktop computer (weighs 11kg, costs \$350)
* a laptop (weighs 5kg, costs \$230), 
* a tablet (weighs 1kg, costs \$115),
* an antique vase (weighs 7kg, costs \$180), 
* a bottle of whisky (weighs 3kg, costs \$75), and 
* a leather jacket (weighs 4kg, costs \$125).

This instance is given in the file `data.pi`.

In [11]:
!cat knapsack/data.pi

instance(Items, Capacity, Values, Weights) =>
    Items = {"tv", "desktop", "laptop", "tablet", "vase", "bottle", "jacket"},
    Capacity = 23,
    Values = {500,350,230,115,180,75,125},
    Weights = {15,11,5,1,7,3,4}.